### PR TITLE
Paddle glu

### LINKED ISSUE #16299: glu

### PR DESCRIPTION
**Pull request by Adeola23**
closes #16299

### CODE REVIEW

**Comment by Adeola23** [comment]
Hi @MuhammadSaeedBatikh I mistakenly closed the PR(https://github.com/unifyai/ivy/pull/17198) you commented on, could pls review the changes suggested

**Comment by Adeola23** [comment]
hi @juliagsy, the failed test are unrelated and for context pls see https://github.com/unifyai/ivy/pull/17198

**Comment by MuhammadSaeedBatikh** [comment]
Hi @Adeola23, the `.filter(lambda x: x[-1] % 2 == 0)` seems to do the trick! However, you only seem to be testing for the last dimension (i.e. axis = -1).

**Comment by MuhammadSaeedBatikh** [comment]
> @MuhammadSaeedBatikh I'm stuck on how to test for axis I'm either getting `ValueError: Expected value range of axis is [-1, 1], but received axis: 1` or hypothesis error

Hi @Adeola23, putting aside this issue for now; the tests are not running properly:
`E   TypeError: _get_input_and_new_shape() missing 1 required positional argument: 'draw'`
You need to add `@st.composite` decorator to `_get_input_and_new_shape` function.
Once added, your tests should run properly:
I got the following, different exception 
```
>           return op(x)
E           RuntimeError: (NotFound) The kernel with key (CPU, NCHW, int8) of kernel `sigmoid` is not registered.
E             [Hint: Expected kernel_iter == iter->second.end() && kernel_key.backend() == Backend::CPU != true, but received kernel_iter == iter->second.end() && kernel_key.backend() == Backend::CPU:1 == true:1.] (at /paddle/paddle/phi/core/kernel_factory.cc:147)
E           
E           Falsifying example: test_paddle_glu(
E               input_x_shape=(['int8'], [array([[0, 0],
E                        [0, 0]], dtype=int8)]),
E               fn_tree='ivy.functional.frontends.paddle.nn.functional.glu',
E               test_flags=FrontendFunctionTestFlags(
E                   num_positional_args=0,
E                   with_out=True,
E                   inplace=False,
E                   as_variable=[False],
E                   native_arrays=[False],
E                   generate_frontend_arrays=False,
E               ),
E               on_device='cpu',
E               frontend='paddle',
E           )
E           
```

**Comment by Adeola23** [comment]
> > @MuhammadSaeedBatikh I'm stuck on how to test for axis I'm either getting `ValueError: Expected value range of axis is [-1, 1], but received axis: 1` or hypothesis error
> 
> Hi @Adeola23, putting aside this issue for now; the tests are not running properly: `E TypeError: _get_input_and_new_shape() missing 1 required positional argument: 'draw'` You need to add `@st.composite` decorator to `_get_input_and_new_shape` function. Once added, your tests should run properly: I got the following, different exception:
> 
> ```
> >           return op(x)
> E           RuntimeError: (NotFound) The kernel with key (CPU, NCHW, int8) of kernel `sigmoid` is not registered.
> E             [Hint: Expected kernel_iter == iter->second.end() && kernel_key.backend() == Backend::CPU != true, but received kernel_iter == iter->second.end() && kernel_key.backend() == Backend::CPU:1 == true:1.] (at /paddle/paddle/phi/core/kernel_factory.cc:147)
> E           
> E           Falsifying example: test_paddle_glu(
> E               input_x_shape=(['int8'], [array([[0, 0],
> E                        [0, 0]], dtype=int8)]),
> E               fn_tree='ivy.functional.frontends.paddle.nn.functional.glu',
> E               test_flags=FrontendFunctionTestFlags(
> E                   num_positional_args=0,
> E                   with_out=True,
> E                   inplace=False,
> E                   as_variable=[False],
> E                   native_arrays=[False],
> E                   generate_frontend_arrays=False,
> E               ),
> E               on_device='cpu',
> E               frontend='paddle',
> E           )
> E           
> ```

Hi @MuhammadSaeedBatikh, I just checked the latest commit and  `@st.composite` decorator is present  and  the test are running fine unless I test for order axis. Pls check the file changed

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_paddle_activation.py [hunk 1]
```diff
@@ -436,6 +437,50 @@ def test_paddle_tanhshrink(
     )
 
 
+@st.composite
```
Comment (Adeola23) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_paddle_activation.py:440]: @MuhammadSaeedBatikh

**Comment by MuhammadSaeedBatikh** [comment]
> > > @MuhammadSaeedBatikh I'm stuck on how to test for axis I'm either getting `ValueError: Expected value range of axis is [-1, 1], but received axis: 1` or hypothesis error
> > 
> > 
> > Hi @Adeola23, putting aside this issue for now; the tests are not running properly: `E TypeError: _get_input_and_new_shape() missing 1 required positional argument: 'draw'` You need to add `@st.composite` decorator to `_get_input_and_new_shape` function. Once added, your tests should run properly: I got the following, different exception:
> > ```
> > >           return op(x)
> > E           RuntimeError: (NotFound) The kernel with key (CPU, NCHW, int8) of kernel `sigmoid` is not registered.
> > E             [Hint: Expected kernel_iter == iter->second.end() && kernel_key.backend() == Backend::CPU != true, but received kernel_iter == iter->second.end() && kernel_key.backend() == Backend::CPU:1 == true:1.] (at /paddle/paddle/phi/core/kernel_factory.cc:147)
> > E           
> > E           Falsifying example: test_paddle_glu(
> > E               input_x_shape=(['int8'], [array([[0, 0],
> > E                        [0, 0]], dtype=int8)]),
> > E               fn_tree='ivy.functional.frontends.paddle.nn.functional.glu',
> > E               test_flags=FrontendFunctionTestFlags(
> > E                   num_positional_args=0,
> > E                   with_out=True,
> > E                   inplace=False,
> > E                   as_variable=[False],
> > E                   native_arrays=[False],
> > E                   generate_frontend_arrays=False,
> > E               ),
> > E               on_device='cpu',
> > E               frontend='paddle',
> > E           )
> > E           
> > ```
> 
> Hi @MuhammadSaeedBatikh, I just checked the latest commit and `@st.composite` decorator is present and the test are running fine unless I test for other axis. Pls check the file changed

You're right. Do you get the same error I posted here?

**Comment by Adeola23** [comment]
> > > > @MuhammadSaeedBatikh I'm stuck on how to test for axis I'm either getting `ValueError: Expected value range of axis is [-1, 1], but received axis: 1` or hypothesis error
> > > 
> > > 
> > > Hi @Adeola23, putting aside this issue for now; the tests are not running properly: `E TypeError: _get_input_and_new_shape() missing 1 required positional argument: 'draw'` You need to add `@st.composite` decorator to `_get_input_and_new_shape` function. Once added, your tests should run properly: I got the following, different exception:
> > > ```
> > > >           return op(x)
> > > E           RuntimeError: (NotFound) The kernel with key (CPU, NCHW, int8) of kernel `sigmoid` is not registered.
> > > E             [Hint: Expected kernel_iter == iter->second.end() && kernel_key.backend() == Backend::CPU != true, but received kernel_iter == iter->second.end() && kernel_key.backend() == Backend::CPU:1 == true:1.] (at /paddle/paddle/phi/core/kernel_factory.cc:147)
> > > E           
> > > E           Falsifying example: test_paddle_glu(
> > > E               input_x_shape=(['int8'], [array([[0, 0],
> > > E                        [0, 0]], dtype=int8)]),
> > > E               fn_tree='ivy.functional.frontends.paddle.nn.functional.glu',
> > > E               test_flags=FrontendFunctionTestFlags(
> > > E                   num_positional_args=0,
> > > E                   with_out=True,
> > > E                   inplace=False,
> > > E                   as_variable=[False],
> > > E                   native_arrays=[False],
> > > E                   generate_frontend_arrays=False,
> > > E               ),
> > > E               on_device='cpu',
> > > E               frontend='paddle',
> > > E           )
> > > E           
> > > ```
> > 
> > 
> > Hi @MuhammadSaeedBatikh, I just checked the latest commit and `@st.composite` decorator is present and the test are running fine unless I test for other axis. Pls check the file changed
> 
> You're right. Do you get the same error I posted here?

Hi @MuhammadSaeedBatikh, I don't get the same error.

**Comment by Adeola23** [comment]
> Hi @Adeola23, the `.filter(lambda x: x[-1] % 2 == 0)` seems to do the trick! However, you only seem to be testing for the last dimension (i.e. axis = -1).

HI @MuhammadSaeedBatik, did you get the same error when you tested here?

**Comment by Adeola23** [comment]
@MuhammadSaeedBatikh

**Comment by MahmoudAshraf97** [comment]
Hi @Adeola23  , are you still working on this function?

File: ivy/functional/frontends/paddle/nn/functional/activation.py [hunk 2]
```diff
@@ -187,6 +187,24 @@ def tanhshrink(
 
 @with_supported_dtypes({"2.4.2 and below": ("float32", "float64")}, "paddle")
 @to_ivy_arrays_and_back
+def glu(x, axis=-1, name=None):
+    y = ivy.shape(x)
```
Comment (MahmoudAshraf97) [inline ivy/functional/frontends/paddle/nn/functional/activation.py:191]: please use a better descriptive variable name other than `y`

File: ivy/functional/frontends/paddle/nn/functional/activation.py [hunk 3]
```diff
@@ -187,6 +187,24 @@ def tanhshrink(
...
+    if y[axis] % 2 != 0:
+        raise ValueError(
+            f"Expected value of shape at axis:{axis} is even, but received odd"
+        )
+    rank = len(ivy.shape(y))
```
Comment (MahmoudAshraf97) [inline ivy/functional/frontends/paddle/nn/functional/activation.py:196]:
```suggestion
    rank = y.ndim
```

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_paddle_activation.py [hunk 4]
```diff
@@ -436,6 +437,50 @@ def test_paddle_tanhshrink(
...
+            shape=new_shape,
+        )
+    )
+    return x_dtype, x
+
+
+@handle_frontend_test(
+    fn_tree="paddle.nn.functional.glu",
+    input_x_shape=_get_input_and_new_shape(),
+    test_with_out=st.just(True),
```
Comment (MahmoudAshraf97) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_paddle_activation.py:463]: why are you forcing the function to test the out argument although it doesn't exist?

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_paddle_activation.py
Comment (MahmoudAshraf97) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_paddle_activation.py:1]: you are not testing for `axis` argument

**Review by MahmoudAshraf97 — CHANGES_REQUESTED** [review]

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_paddle_activation.py
Comment (Adeola23) [reply ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_paddle_activation.py:1]: I'd make the changes and thanks for the response

Comment (Adeola23) [reply ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_paddle_activation.py:1]: Plus I'm going to tag to review the new pull request.

**Pull request closed without merging**